### DISCUSSION TITLE
Mechanism for decorators which consume the function to not prevent redefinitions

### DISCUSSION THREAD

**Discussion by laundmo** [microsoft/pyright#9478] [Ideas]
When a decorator "consumes" the decorated function, by not returning anything (None) therefore replacing the function with None, this could have 2 meanings:
- The decorator SHOULD turn functions into values, and None is a valid value. This is the currently assumed case and pyright recognizes that funcname = None
- The decorator consumes the function (likely storing it internally) and the None returned isn't meant as a useable value. This case seems to be impossible to represent currently in such a way that the same function name can be redefined.

Specifically: I'm running into this problem with a web framework where different HTTP methods are different decorators, and the function name is the html template name. This means I always have to ignore reportRedeclaration errors, even though the previous declaration isn't a function that can be called, since the decorator returns None.

I've tried hinting that the decorator returns NoReturn but this just caused funcname to be of type "NoReturn" (which feels wrong...)

i've no clue if this is something pyright could solve (by understanding NoReturn or Never in this place), or which needs to be solved by the python typing module introducing a new special cased type for this. i just wanted to throw the idea out there,

**Comment by erictraut** [comment]
What your proposing here would need to be codified in the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) if you want pyright (or other standards-compliant Python type checkers) to support it. If you'd like to pitch the idea to the Python typing community, the best place to do that is the [Python typing forum](https://discuss.python.org/c/typing/32).
